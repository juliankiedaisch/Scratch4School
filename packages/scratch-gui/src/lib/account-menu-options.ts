import PropTypes from 'prop-types';

export interface AccountMenuOptions {
    canHaveSession: boolean;

    canRegister?: boolean;
    canLogin?: boolean;
    canLogout?: boolean;

    avatarUrl?: string;
    myStuffUrl?: string;
    profileUrl?: string;
    myClassesUrl?: string;
    myClassUrl?: string;
    accountSettingsUrl?: string;
}

export const AccountMenuOptionsPropTypes = PropTypes.shape({
    canHaveSession: PropTypes.bool.isRequired,

    canRegister: PropTypes.bool,
    canLogin: PropTypes.bool,
    canLogout: PropTypes.bool,

    avatarUrl: PropTypes.string,
    myStuffUrl: PropTypes.string,
    profileUrl: PropTypes.string,
    myClassesUrl: PropTypes.string,
    myClassUrl: PropTypes.string,
    accountSettingsUrl: PropTypes.string
});
