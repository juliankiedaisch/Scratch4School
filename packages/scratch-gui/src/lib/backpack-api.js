import xhr from 'xhr';
import costumePayload from './backpack/costume-payload';
import soundPayload from './backpack/sound-payload';
import spritePayload from './backpack/sprite-payload';
import codePayload from './backpack/code-payload';

// Add a new property for the full thumbnail url, which includes the host
const includeFullUrls = (item, host) => Object.assign({}, item, {
    thumbnailUrl: item.thumbnailUrl || `/backend/api/assets/${item.thumbnail}`,
    bodyUrl: item.bodyUrl || `/backend/api/assets/${item.body}`
});

const getBackpackContents = ({
    host,
    username,
    token,
    limit,
    offset
}) => new Promise((resolve, reject) => {
    xhr({
        method: 'GET',
        uri: `/backend/api/backpack?limit=${limit}&offset=${offset}`,
        headers: {'Authorization': `Bearer ${token}`},
        json: true
    }, (error, response) => {
        if (error || response.statusCode !== 200) {
            return reject(new Error(response.statusCode));
        }
        // Our backend returns an object with 'items' property
        const items = response.body.items || [];
        return resolve(items.map(item => includeFullUrls(item, host)));
    });
});

const saveBackpackObject = ({
    host,
    username,
    token,
    type, // Type of object being saved to the backpack
    mime, // Mime-type of the object being saved
    name, // User-facing name of the object being saved
    body, // Base64-encoded body of the object being saved
    thumbnail // Base64-encoded JPEG thumbnail of the object being saved
}) => new Promise((resolve, reject) => {
    xhr({
        method: 'POST',
        uri: `/backend/api/backpack`,
        headers: {'Authorization': `Bearer ${token}`},
        json: {type, mime, name, body, thumbnail}
    }, (error, response) => {
        if (error || response.statusCode !== 200) {
            return reject(new Error(response.statusCode));
        }
        return resolve(includeFullUrls(response.body, host));
    });
});

const deleteBackpackObject = ({
    host,
    username,
    token,
    id
}) => new Promise((resolve, reject) => {
    xhr({
        method: 'DELETE',
        uri: `/backend/api/backpack/${id}`,
        headers: {'Authorization': `Bearer ${token}`}
    }, (error, response) => {
        if (error || response.statusCode !== 200) {
            return reject(new Error(response.statusCode));
        }
        return resolve(response.body);
    });
});

// Two types of backpack items are not retrievable through storage
// code, as json and sprite3 as arraybuffer zips.
const fetchAs = (responseType, uri) => new Promise((resolve, reject) => {
    xhr({uri, responseType}, (error, response) => {
        if (error || response.statusCode !== 200) {
            return reject(new Error(response.statusCode));
        }
        return resolve(response.body);
    });
});

// These two helpers allow easy fetching of backpack code and sprite zips
const fetchCode = fetchAs.bind(null, 'json');
const fetchSprite = fetchAs.bind(null, 'arraybuffer');

export {
    getBackpackContents,
    saveBackpackObject,
    deleteBackpackObject,
    costumePayload,
    soundPayload,
    spritePayload,
    codePayload,
    fetchCode,
    fetchSprite
};