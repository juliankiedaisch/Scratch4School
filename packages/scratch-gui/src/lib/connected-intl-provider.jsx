import {IntlProvider as ReactIntlProvider} from 'react-intl';
import {connect} from 'react-redux';

const mapStateToProps = state => ({
    key: state.locales.locale,
    locale: state.locales.locale,
    messages: state.locales.messages,
    textComponent: 'span'
});

export default connect(mapStateToProps)(ReactIntlProvider);
