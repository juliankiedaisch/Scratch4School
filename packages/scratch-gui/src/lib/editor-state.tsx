import {createStore, combineReducers, compose, Store} from 'redux';
import localesReducer, {initLocale, localesInitialState} from '../reducers/locales';
import locales from 'scratch-l10n';
import {detectLocale} from './detect-locale';
import {GUIConfig} from '../gui-config';

interface WindowWithDevtools {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: typeof compose;
}

const composeEnhancers = (window as WindowWithDevtools).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

// TypeScript doesn't know about require here, and we don't want to change behavior, so...
declare function require(path: '../reducers/gui'): typeof import('../reducers/gui');
declare function require(path: 'scratch-paint'): typeof import('scratch-paint');
declare function require(path: '../legacy-config'): typeof import('../legacy-config');

export interface EditorStateParams {
    localesOnly?: boolean;
    isFullScreen?: boolean;
    isPlayerOnly?: boolean;
    showTelemetryModal?: boolean;
    isEmbedded?: boolean;
}

/**
 * Manages an editor's Redux state.
 *
 * To be used in tandem with an AppStateHOC component to be provided to the editor.
 */
export class EditorState {
    /**
     * The redux store that this class wraps.
     */
    public readonly store: Store<unknown>;

    constructor (params: EditorStateParams, configFactory: () => GUIConfig) {
        let initialState = {};
        let reducers = {};
        let enhancer;

        let initializedLocales = localesInitialState;
        const locale = detectLocale(Object.keys(locales));
        if (locale !== 'en') {
            initializedLocales = initLocale(initializedLocales, locale);
        }
        if (params.localesOnly) {
            // Used for instantiating minimal state for the unsupported
            // browser modal
            reducers = {locales: localesReducer};
            initialState = {locales: initializedLocales};
            enhancer = composeEnhancers();
        } else {
            // You are right, this is gross. But it's necessary to avoid
            // importing unneeded code that will crash unsupported browsers.
            const guiRedux = require('../reducers/gui');
            const guiReducer = guiRedux.default;
            const {
                buildInitialState,
                guiMiddleware,
                initFullScreen,
                initPlayer,
                initTelemetryModal,
                initEmbedded
            } = guiRedux;
            const {ScratchPaintReducer} = require('scratch-paint');

            const configOrLegacy = configFactory ?
                configFactory() :
                require('../legacy-config').legacyConfig;

            let initializedGui = buildInitialState(configOrLegacy);
            if (params.isFullScreen || params.isPlayerOnly || params.isEmbedded) {
                if (params.isFullScreen) {
                    initializedGui = initFullScreen(initializedGui);
                }
                if (params.isPlayerOnly) {
                    initializedGui = initPlayer(initializedGui);
                }
                if (params.isEmbedded) {
                    initializedGui = initEmbedded(initializedGui);
                }
            } else if (params.showTelemetryModal) {
                initializedGui = initTelemetryModal(initializedGui);
            }
            reducers = {
                locales: localesReducer,
                scratchGui: guiReducer,
                scratchPaint: ScratchPaintReducer
            };
            initialState = {
                locales: initializedLocales,
                scratchGui: initializedGui
            };
            enhancer = composeEnhancers(guiMiddleware);
        }
        const reducer = combineReducers(reducers);
        this.store = createStore(reducer, initialState, enhancer);
    }

    dispatch (action) {
        this.store.dispatch(action);
    }
}
