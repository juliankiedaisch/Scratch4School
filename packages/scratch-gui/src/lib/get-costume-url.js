import {inlineSvgFonts} from '@scratch/scratch-svg-renderer';

// Contains 'font-family', but doesn't only contain 'font-family="none"'
const HAS_FONT_REGEXP = 'font-family(?!="none")';

const getCostumeUrl = (function () {
    // The caching here assumes the scratchStorage instance doesn't change.
    // Which it shouldn't but it's good to keep that in mind anyway.
    let cachedAssetId;
    let cachedUrl;

    return function (scratchStorage, asset) {

        if (cachedAssetId === asset.assetId) {
            return cachedUrl;
        }

        cachedAssetId = asset.assetId;

        // If the SVG refers to fonts, they must be inlined in order to display correctly in the img tag.
        // Avoid parsing the SVG when possible, since it's expensive.
        if (asset.assetType === scratchStorage.AssetType.ImageVector) {
            const svgString = asset.decodeText();
            if (svgString.match(HAS_FONT_REGEXP)) {
                const svgText = inlineSvgFonts(svgString);
                cachedUrl = `data:image/svg+xml;utf8,${encodeURIComponent(svgText)}`;
            } else {
                cachedUrl = asset.encodeDataURI();
            }
        } else {
            cachedUrl = asset.encodeDataURI();
        }

        return cachedUrl;
    };
}());

export {
    getCostumeUrl as default,
    HAS_FONT_REGEXP
};
