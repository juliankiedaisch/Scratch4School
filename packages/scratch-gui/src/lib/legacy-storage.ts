import {ScratchStorage, Asset} from 'scratch-storage';

import defaultProject from './default-project';
import {GUIStorage, TranslatorFunction} from '../gui-config';

import saveProjectToServer from '../lib/save-project-to-server';

export class LegacyStorage implements GUIStorage {
    private projectHost?: string;
    private projectToken?: string;
    private assetHost?: string;
    private backpackHost?: string;
    private translator?: TranslatorFunction;

    readonly scratchStorage = new ScratchStorage();

    constructor () {
        this.cacheDefaultProject(this.scratchStorage);
        this.addOfficialScratchWebStores(this.scratchStorage);
    }

    setProjectHost (host: string): void {
        this.projectHost = host;
    }

    setProjectToken (token: string): void {
        this.projectToken = token;
    }

    setProjectMetadata (projectId: string | null | undefined): void {
        const {RequestMetadata, setMetadata, unsetMetadata} = this.scratchStorage.scratchFetch;

        // If project ID is '0' or zero, it's not a real project ID. In that case, remove the project ID metadata.
        // Same if it's null undefined.
        if (projectId && projectId !== '0') {
            setMetadata(RequestMetadata.ProjectId, projectId);
        } else {
            unsetMetadata(RequestMetadata.ProjectId);
        }
    }

    setAssetHost (host: string): void {
        this.assetHost = host;
    }

    setTranslatorFunction (translator: TranslatorFunction): void {
        this.translator = translator;

        // TODO: Verify that this is correct
        this.cacheDefaultProject(this.scratchStorage);
    }

    setBackpackHost (host: string): void {
        const shouldAddSource = !this.backpackHost;
        if (shouldAddSource) {
            const AssetType = this.scratchStorage.AssetType;

            this.scratchStorage.addWebStore(
                [AssetType.ImageVector, AssetType.ImageBitmap, AssetType.Sound],
                this.getBackpackAssetURL.bind(this)
            );
        }

        this.backpackHost = host;
    }

    saveProject (
        projectId: number,
        vmState: string,
        params: { originalId: string; isCopy: boolean; isRemix: boolean; title: string; }
    ): Promise<{ id: string | number; }> {
        if (!this.projectHost) {
            return Promise.reject(new Error('Project host not set'));
        }
        // Haven't inlined the code here so that we can keep Git history on the implementation, just in case
        return saveProjectToServer(this.projectHost, projectId, vmState, params);
    }

    saveProjectThumbnail = function (projectId, dataURL) {
    // Log that we're attempting to save a thumbnail
    console.info(`Saving thumbnail for project ${projectId}`);
    
    return new Promise((resolve, reject) => {
        if (!projectId) {
            reject(new Error('Cannot save thumbnail without a project ID'));
            return;
        }
        
        // Make sure we have a session ID
        const sessionId = localStorage.getItem('session_id');
        if (!sessionId) {
            reject(new Error('No session ID found, cannot save thumbnail'));
            return;
        }
        
        // Safely handle dataURL - check if it's a string and has the expected format
        let thumbnailBlob;
        
        try {
            if (typeof dataURL === 'string' && dataURL.startsWith('data:')) {
                // Handle data URL string
                const parts = dataURL.split(',');
                if (parts.length !== 2) {
                    throw new Error('Invalid data URL format');
                }
                
                const byteString = atob(parts[1]);
                const mimeString = parts[0].split(':')[1].split(';')[0];
                
                const arrayBuffer = new ArrayBuffer(byteString.length);
                const intArray = new Uint8Array(arrayBuffer);
                
                for (let i = 0; i < byteString.length; i++) {
                    intArray[i] = byteString.charCodeAt(i);
                }
                
                thumbnailBlob = new Blob([arrayBuffer], {type: mimeString});
            } else if (dataURL instanceof HTMLCanvasElement) {
                // Handle canvas element
                console.info('Converting canvas to blob for thumbnail');
                return new Promise(resolve => {
                    dataURL.toBlob(blob => {
                        resolve(blob);
                    }, 'image/png');
                })
                .then(blob => {
                    if (blob) uploadThumbnailBlob(blob);
                    else resolve({});
                })
                .catch(err => {
                    console.warn('Error converting canvas to blob:', err);
                    resolve({}); // Resolve anyway to not block the save process
                });
            } else {
                // Create a default thumbnail if dataURL is invalid
                console.warn('Invalid thumbnail data, creating default thumbnail');
                const canvas = document.createElement('canvas');
                canvas.width = 480;
                canvas.height = 360;
                const ctx = canvas.getContext('2d');
                
                // TypeScript null check - ensure ctx is not null before using it
                if (ctx) {
                    ctx.fillStyle = '#159cff'; // Scratch blue
                    ctx.fillRect(0, 0, 480, 360);
                    ctx.fillStyle = 'white';
                    ctx.font = 'bold 36px sans-serif';
                    ctx.textAlign = 'center';
                    ctx.fillText('Project', 240, 180);
                    
                    return new Promise(resolve => {
                        canvas.toBlob(blob => {
                            resolve(blob);
                        }, 'image/png');
                    })
                    .then(blob => {
                        if (blob) uploadThumbnailBlob(blob);
                        else resolve({});
                    })
                    .catch(err => {
                        console.warn('Error creating default thumbnail:', err);
                        resolve({}); // Resolve anyway to not block the save process
                    });
                } else {
                    // Canvas 2D context not available
                    console.warn('Canvas 2D context not available');
                    resolve({});
                    return;
                }
            }
        } catch (e) {
            console.warn('Error processing thumbnail data:', e);
            resolve({}); // Resolve anyway to not block the save process
            return;
        }
        
        // If we reach this point, we have a valid thumbnailBlob
        if (thumbnailBlob) {
            uploadThumbnailBlob(thumbnailBlob);
        } else {
            resolve({}); // Resolve with empty object if no thumbnail blob
        }
        
        // Helper function to upload the thumbnail blob
        function uploadThumbnailBlob(blob) {
            // Create a FormData object and append the file
            const formData = new FormData();
            formData.append('thumbnail', blob, 'thumbnail.png');
            
            // Use the correct URL pattern with /backend prefix
            const url = `/backend/api/projects/${projectId}/thumbnail`;
            
            // Fix the type error: sessionId is string | null, but we've checked it's not null above
            // Use type assertion to tell TypeScript that sessionId is definitely a string here
            const headers: HeadersInit = {
                'X-Session-ID': sessionId as string // Type assertion fixes the error
            };
            
            fetch(url, {
                method: 'POST',
                headers,
                body: formData,
                credentials: 'include'
            })
            .then(response => {
                if (!response.ok) {
                    console.warn(`Thumbnail upload failed with status: ${response.status}`);
                    return response.text().then(text => {
                        throw new Error(`Thumbnail upload failed: ${text}`);
                    });
                }
                return response.json();
            })
            .then(body => {
                console.info('Thumbnail upload successful');
                resolve(body);
            })
            .catch(err => {
                console.warn('Error uploading thumbnail:', err);
                // Resolve anyway since thumbnail upload isn't critical
                resolve({});
            });
        }
    });
};

    private cacheDefaultProject (storage: ScratchStorage) {
        const defaultProjectAssets = defaultProject(this.translator);
        defaultProjectAssets.forEach(asset => storage.builtinHelper._store(
            storage.AssetType[asset.assetType],
            storage.DataFormat[asset.dataFormat],
            asset.data,
            asset.id
        ));
    }

    private addOfficialScratchWebStores (storage: ScratchStorage) {
        storage.addWebStore(
            [storage.AssetType.Project],
            this.getProjectGetConfig.bind(this),
            this.getProjectCreateConfig.bind(this),
            this.getProjectUpdateConfig.bind(this)
        );

        storage.addWebStore(
            [storage.AssetType.ImageVector, storage.AssetType.ImageBitmap, storage.AssetType.Sound],
            this.getAssetGetConfig.bind(this),
            // We set both the create and update configs to the same method because
            // storage assumes it should update if there is an assetId, but the
            // asset store uses the assetId as part of the create URI.
            this.getAssetCreateConfig.bind(this),
            this.getAssetCreateConfig.bind(this)
        );

        storage.addWebStore(
            [storage.AssetType.Sound],
            asset => `static/extension-assets/scratch3_music/${asset.assetId}.${asset.dataFormat}`
        );

        storage.addWebStore(
            [storage.AssetType.ImageVector, storage.AssetType.ImageBitmap, storage.AssetType.Sound],
            this.getAssetGetConfigBackend.bind(this),
            // We set both the create and update configs to the same method because
            // storage assumes it should update if there is an assetId, but the
            // asset store uses the assetId as part of the create URI.
            this.getAssetCreateConfig.bind(this),
            this.getAssetCreateConfig.bind(this)
        );

    }


    private getProjectGetConfig (projectAsset) {
        const path = `${this.projectHost}/${projectAsset.assetId}`;
        const qs = this.projectToken ? `?token=${this.projectToken}` : '';
        return path + qs;
    }

    private getProjectCreateConfig () {
        return {
            url: `${this.projectHost}/`,
            withCredentials: true
        };
    }

    private getProjectUpdateConfig (projectAsset: Asset) {
        return {
            url: `${this.projectHost}/${projectAsset.assetId}`,
            withCredentials: true
        };
    }

    private getAssetGetConfigBackend (asset: Asset) {
        return `/backend/api/assets/${asset.assetId}`;
    }

    private getAssetGetConfig (asset: Asset) {
        return `${this.assetHost}/internalapi/asset/${asset.assetId}.${asset.dataFormat}/get/`;
    }
    private getAssetCreateConfig (asset: Asset) {
        return {
            // There is no such thing as updating assets, but storage assumes it
            // should update if there is an assetId, and the asset store uses the
            // assetId as part of the create URI. So, force the method to POST.
            // Then when storage finds this config to use for the "update", still POSTs
            method: 'post',
            url: `${this.assetHost}/${asset.assetId}.${asset.dataFormat}`,
            withCredentials: true
        };
    }

    private getBackpackAssetURL (asset) {
        return `${this.backpackHost}/${asset.assetId}.${asset.dataFormat}`;
    }

}
