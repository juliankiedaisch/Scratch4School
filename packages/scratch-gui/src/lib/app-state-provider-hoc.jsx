import React from 'react';
import {Provider} from 'react-redux';
import PropTypes from 'prop-types';

import {EditorState} from './editor-state';
import {setPlayer, setFullScreen, setEmbedded} from '../reducers/mode.js';
import ConnectedIntlProvider from './connected-intl-provider.jsx';

/**
 * Wraps the editor into the redux state contained within an EditorState instance.
 *
 * @param {React.Component} WrappedComponent - component to provide state for
 *
 * @returns {React.Component} component with redux and intl state provided
 */
export const AppStateProviderHOC = function (WrappedComponent) {
    class AppStateWrapper extends React.Component {
        componentDidUpdate (prevProps) {
            if (this.props.localesOnly) return;
            if (prevProps.isPlayerOnly !== this.props.isPlayerOnly) {
                this.props.appState.store.dispatch(setPlayer(this.props.isPlayerOnly));
            }
            if (prevProps.isFullScreen !== this.props.isFullScreen) {
                this.props.appState.store.dispatch(setFullScreen(this.props.isFullScreen));
            }
            if (prevProps.isEmbedded !== this.props.isEmbedded) {
                this.props.appState.store.dispatch(setEmbedded(this.props.isEmbedded));
            }
        }

        render () {
            const {
                appState,
                isFullScreen, // eslint-disable-line no-unused-vars
                isPlayerOnly, // eslint-disable-line no-unused-vars
                showTelemetryModal, // eslint-disable-line no-unused-vars
                isEmbedded, // eslint-disable-line no-unused-vars
                ...componentProps
            } = this.props;
            return (
                <Provider store={appState.store}>
                    <ConnectedIntlProvider>
                        <WrappedComponent
                            {...componentProps}
                        />
                    </ConnectedIntlProvider>
                </Provider>
            );
        }
    }
    AppStateWrapper.propTypes = {
        appState: PropTypes.instanceOf(EditorState),
        localesOnly: PropTypes.bool,
        isFullScreen: PropTypes.bool,
        isPlayerOnly: PropTypes.bool,
        isTelemetryEnabled: PropTypes.bool,
        showTelemetryModal: PropTypes.bool,
        isEmbedded: PropTypes.bool
    };
    return AppStateWrapper;
};
