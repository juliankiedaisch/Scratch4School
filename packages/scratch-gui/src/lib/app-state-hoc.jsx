import React from 'react';
import PropTypes from 'prop-types';

import {EditorState} from './editor-state';
import {AppStateProviderHOC} from './app-state-provider-hoc';
// Import the UserProvider from our context
import {UserProvider} from '../contexts/UserContext';

/**
 * Higher Order Component to provide redux state. If an `intl` prop is provided
 * it will override the internal `intl` redux state
 *
 * @param {React.Component} WrappedComponent - component to provide state for
 * @param {boolean} localesOnly - only provide the locale state, not everything
 *                      required by the GUI. Used to exclude excess state when
 *                      only rendering modals, not the GUI.
 * @param {GUIConfigFactory} configFactory - The configuration to use.
 *
 * @returns {React.Component} component with redux and intl state provided
 */
const AppStateHOC = function (WrappedComponent, localesOnly, configFactory) {
    const AppStateProvider = AppStateProviderHOC(WrappedComponent);

    class AppStateWrapper extends React.Component {
        constructor (props) {
            super(props);

            this.appState = new EditorState({
                localesOnly,
                isFullScreen: props.isFullScreen,
                isPlayerOnly: props.isPlayerOnly,
                showTelemetryModal: props.showTelemetryModal,
                isEmbedded: props.isEmbedded
            }, configFactory);
        }

        render () {
            return (
                // Wrap the AppStateProvider with UserProvider to enable OAuth authentication
                <UserProvider>
                    <AppStateProvider
                        appState={this.appState}
                        localesOnly={localesOnly}
                        {...this.props}
                    />
                </UserProvider>
            );
        }
    }
    AppStateWrapper.propTypes = {
        isFullScreen: PropTypes.bool,
        isPlayerOnly: PropTypes.bool,
        isTelemetryEnabled: PropTypes.bool,
        showTelemetryModal: PropTypes.bool,
        isEmbedded: PropTypes.bool
    };
    return AppStateWrapper;
};

export default AppStateHOC;