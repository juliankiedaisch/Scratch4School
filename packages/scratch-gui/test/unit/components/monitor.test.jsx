import React from 'react';
import {render} from '@testing-library/react';
import Monitor from '../../../src/components/monitor/monitor';
import {DARK_THEME, DEFAULT_THEME} from '../../../src/lib/themes';

jest.mock('../../../src/lib/themes/default');
jest.mock('../../../src/lib/themes/dark');

describe('Monitor Component', () => {
    const noop = jest.fn();

    const defaultProps = {
        category: 'motion',
         
        componentRef: noop,
        draggable: false,
        label: 'My label',
        mode: 'default',
         
        onDragEnd: noop,
         
        onNextMode: noop
    };

    test('it selects the correct colors based on default theme', () => {
        const {container} = render(<Monitor
            {...defaultProps}
            theme={DEFAULT_THEME}
        />);

        expect(container.firstChild).toMatchSnapshot();
    });

    test('it selects the correct colors based on dark mode theme', () => {
        const {container} = render(<Monitor
            {...defaultProps}
            theme={DARK_THEME}
        />);

        expect(container.firstChild).toMatchSnapshot();
    });
});
