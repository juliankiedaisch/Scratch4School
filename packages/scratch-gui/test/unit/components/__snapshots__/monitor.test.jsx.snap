// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Monitor Component it selects the correct colors based on dark mode theme 1`] = `
<div
  class="react-draggable"
  style="transform: translate(0px,0px);"
>
  <span
    class="ContextMenuTrigger"
    data-state="closed"
  >
    <div>
      <div>
        <div>
          My label
        </div>
        <div
          style="background: rgb(170, 170, 170); color: rgb(187, 187, 187);"
        />
      </div>
    </div>
  </span>
</div>
`;

exports[`Monitor Component it selects the correct colors based on default theme 1`] = `
<div
  class="react-draggable"
  style="transform: translate(0px,0px);"
>
  <span
    class="ContextMenuTrigger"
    data-state="closed"
  >
    <div>
      <div>
        <div>
          My label
        </div>
        <div
          style="background: rgb(17, 17, 17); color: rgb(68, 68, 68);"
        />
      </div>
    </div>
  </span>
</div>
`;
